### ISSUE TITLE
json document with object array

### ISSUE THREAD

**Issue by dzhang** [typesense/typesense#137]
Hi

does typesense support json document with object array, such as blog post with comments

{"title": "title", "post": "content", "comments": [{"id": "1"},{"id": "2"}]}

Thanks
Dong

**Comment by kishorenc** [comment]
👋 @dzhang 

Are you looking to query on the `comments.id` values? Typesense does not allow you to index and search on such nested values. However, you can store a document with nested fields and retrieve them as non-indexed fields for display.

**Comment by HananoshikaYomaru** [comment]
can you explain more about this `However, you can store a document with nested fields and retrieve them as non-indexed fields for display.`

I cannot find any documentation about this in the web.

**Comment by kishorenc** [comment]
@HananoshikaYomaru 

When you send Typesense a document, Typesense stores that document on-disk as it is. Let's take a sample document:

```
{
   "person": {
      "name": "Jack",
      "age": 25
   },
   "planet": "Earth"
}
```

While Typesense cannot index the `person.name` and `person.age` nested fields directly for searching/filtering operations, the whole document will be stored verbatim and so the nested fields will be available in the document that is present in the "hits" response during searching.

**Comment by liemle3893** [comment]
Hi there, same issue here. Its was closed without any further explain so I want to reopen this issue.

I cannot put document which contains array of objects, eg: `{"title": "title", "post": "content", "arr_obj": [{"id": "1"},{"id": "2"}]}`

I got following error: `"error":"Type of field `arr_obj` is invalid."`

I'm not trying to search by that `arr_obj.id`, I just want to store the whole document.

**Comment by kishorenc** [comment]
Can you please post the full collection schema and a sample document that reproduces the issue?

**Comment by liemle3893** [comment]
Collection definition:
```
{
  "name": "example",
  "fields": [
    {
      "name": ".*",
      "type": "auto"
    }
  ]
}
```

Document:

book.jsonl
```
{"title":"The Hunger Games","authors":["Suzanne Collins"],"publication_year":2008,"id":"1","average_rating":4.34,"image_url":"https://images.gr-assets.com/books/1447303603m/2767052.jpg","ratings_count":4780653, "rating":{"count":1,"wtf":"111"},"arr_obj":[{"id":1}]}
```

Docker:
```
docker run -p 8108:8108 -v /tmp/typesense/typesense-data:/data typesense/typesense:0.23.0 --data-dir /data --api-key=Hu52dwsas2AdxdE
```
Environment:
```
export TYPESENSE_HOST='http://localhost:8108'
export TYPESENSE_API_KEY=Hu52dwsas2AdxdE
```

Import
```
curl "${TYPESENSE_HOST}/collections/example/documents/import?action=upsert" \
      -X POST \
      -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" \
      --data-binary @./books.jsonl
```

**Comment by liemle3893** [comment]
Oh, yeah. I figured it out.

Change to following collection definition should do the trick:

```
{
  "name": "example",
  "fields": [
    {
      "name": ".*",
      "type": "auto"
    },
    // Array cannot be indexed, so we must be excluded from beginning.
    {
      "name": "arr_obj",
      "type": "auto",
      "index": false
    }
  ]
}
```

Tks @kishorenc  for your fast response.

**Comment by kishorenc** [comment]
@liemle3893 Yes, ideally Typesense should not require an `index: false` here just as you don't need for regular object fields, but this is a work around for now. In the next version, we will support indexing of nested structures, so this behavior will not occur.

**Comment by wennals** [comment]
@kishorenc  Do you know when this functionality will be released ?

**Comment by Plinpod** [comment]
When will indexing of nested structures be made available?

**Comment by jasonbosco** [comment]
We're tracking this in #227. 

The feature is already available in the latest RC build `0.24.0.rcn42`, so you can use it now. Here's how to use it: https://github.com/typesense/typesense/issues/227#issuecomment-1211666924
